### ISSUE TITLE
TypeVar inference on default arguments

### ISSUE THREAD

**Issue by n-gao** [astral-sh/ty#3609]
### Summary

The following code passes successfully under pyright
```py
from typing import Callable


class A(int): ...


def f[T](a: int, b: Callable[[int], T] = A) -> T:
    return b(a)

```
where if `b` is not specified the return value is inferred to be `A`. ty gives the following:
```
error[invalid-parameter-default]: Default value of type `def g(x: int) -> str` is not assignable to annotated parameter type `(int, /) -> T@f`
 --> test.py:8:18
  |
8 | def f[T](a: int, b: Callable[[int], T] = g) -> T:
  |                  ^^^^^^^^^^^^^^^^^^^^^^^^^
  |

Found 1 diagnostic
```

**Comment by AlexWaygood** [comment]
Thanks for the report. I think this is the same as https://github.com/astral-sh/ty/issues/592
